src/libcore/ptr/mod.rs

```diff
@@ -123,6 +123,8 @@ mod mut_ptr;
 ///
 /// * `to_drop` must be properly aligned.
 ///
+/// * `to_drop` must point to a properly initialized value of type `T`.
```

Comment:
Hm... this one is tricky, actually. Namely, in general, this is not sufficient... the term "initialized" here, in the sense it is [used in these docs](https://doc.rust-lang.org/nightly/std/mem/union.MaybeUninit.html#method.assume_init), is basically what I called ["validity invariant"](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html); however, for `drop_in_place`, *in general* we even need the "safety invariant" to hold. Except that for some types we know that dropping them doesn't do anything, so e.g. for `ManuallyDrop<Vec<u8>>`, we do *not* need the safety invariant to hold.

I think I'd be more comfortable with something that points out that, since `T`'s destructor is going to be called on the value, UB may be caused if the value is not actually valid for destructing -- the details of this depend on the type.

Cc @Centril @rust-lang/wg-unsafe-code-guidelines

Reply:
This sounds quite reasonable to me.

Reply:
I would leave this bullet there, since this first produces a `&mut T` to call `drop` on, so if it isn't initialized this is inst-UB (like in all the other methods, which return a `T`).

Then we can add something like:
```
* The `T` value `to_drop` points to must be valid for destructing, which may mean additional invariants must hold for its fields - this is type dependent.
```
Is this sufficient? Is there a link for further reading which would be relevant?

Reply:
@RalfJung Hmm, reading through the post you linked to, you're stating that (at least currently) having a `&mut T` to a non-fully-initialized `T` is actually Ok. Does this mean we should remove that bullet about "properly initialized", since, depending on the destructor, we might e.g. not be reading some field, making it fine for it to be uninitialized?

Reply:
> I would leave this bullet there, since this first produces a &mut T to call drop on, so if it isn't initialized this is inst-UB (like in all the other methods, which return a T).

Arguably it wouldn't do that for types that don't even need dropping, would it?

Though, conservatively, it is probably best to declare e.g. `drop_in_place(0 as *mut i32)` UB for now even if we want to relax this later.

> "The T value to_drop points to must be valid for destructing, which may mean it must uphold additional invariants - this is type dependent."

Yeah, adding something like this would be better.
I don't know of a helpful link here.

Reply:
> Arguably it wouldn't do that for types that don't even need dropping, would it?

This actually made me think of another requirement, at least in the case of types with a `drop` implementation: the pointer must not alias.

Reply:
Well, must not alias *what*?

It must be a valid mutable reference, basically.

Reply:
Well, it must not alias any other reference in the program. Perhaps the correct wording is "valid for writes" (as appearing elsewhere in this module; the current docs say "valid for reads", which isn't sufficient.

Reply:
Anyway, it seems to me that a type *not* having to drop isn't something you can rely on, so I would say it's better make this UB for any type. Or otherwise - if you know that this type doesn't need dropping (e.g. this is an integer primitive), you shouldn't need to call `drop_in_place` on it anyway.

Reply:
You should be able to rely on it for `T: Copy`.

Reply:
> Well, it must not alias any other reference in the program. 

That is, at best, really imprecise:
```rust
let x = &mut <...>;
let y = &mut *x;
drop_in_place(y);
```
That program is fine even though `x` and `y` alias.

I suggest we don't try to start explaining the aliasing rules here, which we also avoid everywhere else in the docs as it is really, really hard.

Reply:
> That is, at best, really imprecise:

Right, right. That's why I said, we can write "valid for writes" and leave it at that.

> You should be able to rely on it for `T: Copy`.

So perhaps, leave the current requirements as is (alignment+valid for reads), and then, "In addition, if `T` isn't `Copy`", valid for writes+ready for destructing (type specific)?

Arguably, if `T` is `Copy`, even reads and alignment are unnecessary: currently looking at the generated LLVM IR (for calling `to_drop` on `*mut usize`), the pointer is passed around but never actually read into.

Reply:
> Right, right. That's why I said, we can write "valid for writes" and leave it at that.

:+1:

> So perhaps, leave the current requirements as is (alignment+valid for reads), and then, "In addition, if T isn't Copy", valid for writes+ready for destructing (type specific)?

"valid for reads" doesn't make much sense here, though... not sure why that's even in the list. But the conservative thing would be to require "valid for writes" *always*, plus type-specific destructing things.

Reply:
Well, a `&mut T` must also be valid for reads, I assume? I updated with the results of this discussion.

Reply:
"valid for writes" implies "valid for reads"

Reply:
"both reads and writes" is fine, though

Reply:
For anyone who stubles upon this: it seems that relying on a type not being `Drop` if it's `Copy` is not necessarily wished for (AFAII, related to making `UnsafeCell` impl `Copy`):
https://github.com/rust-lang/rust/issues/60715#issuecomment-515739427

Reply:
I think you misunderstood that other post... even when `Cell: Copy`, I doubt we'll ever allow types that are `Copy + Drop`. Since `Copy` types are implicitly duplicated all the time, all of these copies would have to have their destructor running -- that is the kind of implicit cost Rust tries hard to avoid.

Reply:
Well, it does directly say

> People already abuse it for making sure a type has a no-op Drop impl, which is its own headache [..]

But this might of course be a misguided complaint?

Reply:
I think @pythonesque just wanted to say that saying `T: Copy` is a rather roundabout and bad way to say "T has a no-op Drop". It is *correct*, but (a) the intent is unclear (it looks like you want to copy the type but really you care about its lack of a drop shim), and (b) it is imprecise (many types, such as `&mut T`, have a no-op Drop but are not `Copy`).

Reply:
Ok, thanks for the explanation!